espnet2/bin/asr_align.py

```diff
@@ -0,0 +1,675 @@
+            index_duration = self.samples_to_frames_ratio / self.fs
+        elif self.time_stamps == "auto":
+            samples_to_frames_ratio = speech_len / lpz_len
+            index_duration = samples_to_frames_ratio / self.fs
+        timing_cfg["index_duration"] = index_duration
+        return timing_cfg
+
+    def estimate_samples_to_frames_ratio(self, speech_len=215040):
+        """Determine the ratio of encoded frames to sample points.
+
+        This method helps to determine the time a single encoded frame occupies.
+        As the sample rate already gave the number of samples, only the ratio
+        of samples per encoded CTC frame are needed. This function estimates them by
+        doing one inference, which is only needed once.
+
+        Args:
+            speech_len: Length of randomly generated speech vector for single
+                inference. Default: 215040=(2048 * 3 * 5 * 7).
+
+        Returns:
+            samples_to_frames_ratio: Estimated ratio.
+        """
+        random_input = torch.rand(speech_len)
+        lpz = self.get_lpz(random_input)
+        lpz_len = lpz.shape[0]
+        # Most frontends (DefaultFrontend, SlidingWindow) discard trailing data
+        lpz_len = lpz_len + 1
+        samples_to_frames_ratio = speech_len // lpz_len
+        return samples_to_frames_ratio
+
+    @torch.no_grad()
+    def get_lpz(self, speech: torch.Tensor):
+        """Obtain CTC posterior log probabilities for given speech data."""
+        # data: (Nsamples,) -> (1, Nsamples)
+        speech = speech.unsqueeze(0).to(getattr(torch, self.dtype))
+        # lenghts: (1,)
+        lengths = speech.new_full([1], dtype=torch.long, fill_value=speech.size(1))
+        batch = {"speech": speech, "speech_lengths": lengths}
+        batch = to_device(batch, device=self.device)
+        # Encode input
+        enc, _ = self.asr_model.encode(**batch)
+        assert len(enc) == 1, len(enc)
+        # Apply ctc layer to obtain log character probabilities
+        lpz = self.ctc.log_softmax(enc.unsqueeze(0)).detach()
+        #  Shape should be ( <time steps>, <classes> )
+        lpz = lpz.squeeze(0).squeeze(0).cpu().numpy()
+        return lpz
+
+    def _text_preprocess(self, text):
+        """Cleanup text and extract utterance IDs."""
+        utt_ids = None
+        # Handle multiline strings
+        if isinstance(text, str):
+            text = text.splitlines()
+        # Remove empty lines
+        text = list(filter(len, text))
+        # Handle kaldi-style text format
+        if self.kaldi_style_text:
+            utt_ids = [utt.split(" ", 1)[0] for utt in text]
+            text = [utt.split(" ", 1)[1] for utt in text]
+        return utt_ids, text
+
+    def get_segments(self, text, lpz, timing_cfg=None):
+        """Obtain segments for given utterance texts and CTC log posteriors."""
+        config = self.config
+        # text is needed in the form of a list.
+        utt_ids, text = self._text_preprocess(text)
```

Comment:
The preprocess must be done by preprocessor class. Could you refer the following ?https://github.com/espnet/espnet/blob/967deef41e579af031ba5d96774f2ec74d250d89/espnet2/bin/tts_inference.py#L141

Reply:
Sorry, I'm not sure  what text format you used here.

ESPnet uses the following text format as here:

```
uttid1 hello world
uttid2 abc der
```

What does "tokenized_text" mean here?

Reply:
In my understanding, you are converting a raw text to integer ids: e.g `hello world` -> `3 10`. If so,
 we must use preprocess_fn. It's not allowed to refer `asr_model.token_list` directly because no one knows how to convert a  raw text to integer ids except for the preprocessor class.

Then, the option of "tokenized_text" is not needed because the input type is always raw text in espnet2.

Reply:
Thank you for the hint! `tokenized_text` was a workaround, it's not needed anymore when we have `preprocess_fn`.

So the input to the module shall be:
```python
text = """uttid1 hello world
uttid2 abc def
"""
```
Then, my module would split the utterances like this:
```python
text = ["hello world", "abc def"]
utt_ids = ["uttid1", "uttid2"]
```
I could then use `preprocess_fn` from `ASRTask.build_preprocess_fn` to obtain the token IDs, maybe like this:
```python
text = [preprocess_fn(utt) for utt in text]
# [np.array([ 3, 10]), np.array([ 17, 19])]
```
Will this work? I'll have to rewrite some functions and test, takes a few days...

Reply:
> Will this work? I'll have to rewrite some functions and test, takes a few days...

Yes, thanks!

Reply:
The model now uses `preprocess_fn` and the kaldi style text format by default.

Reply:
The alignment result is a bit different when we use the `preprocess_fn`. The preprocessing method we used in espnet 1 scans for partial tokens, which is better for models with long tokens in their dictionaries. Therefore, I additionally included an option to handle the input text as in espnet 1 (with `text_converter="classic"`)...

The `asr_model.token_list` is used to generate debug information. For example, the `segments.state_list` of the `ctc_align_test.wav`:

```python
"".join(segments.state_list)
# "εεεεεεεεεεεεεεεε<blank>THEε<space>SεALEε<space>OεF<space>THE<space>HεOTεεEεLεεεεεSε<blank>εεεεIS<space>εPAεRT<space>εOF<space>HεOLεIεDAY'S<space>εSTRεATεEεGεYε<blank>εεεTO<space>SεELεL<space>OFεF<space>εAεSSεεεETεSε<blank>ANεD<space>εCOεNεCEεNεTRεATEε<space>ONε<space>PεROPεERεTYε<space>εMεANεAεGEεMEεNεT<blank>"
```

Reply:
> The preprocessing method we used in espnet 1 scans for partial tokens, which is better for models with long tokens in their dictionaries. 

Sorry, I couldn't understand totally. Could you explain a bit more?

Reply:
Let's say we have a text `text = ["cat"]` and a dictionary that includes the word cat as well as its parts: `char_list = ["•", "UNK", "a", "c", "t", "cat"]`.
The "tokenize" method that uses the `preprocess_fn` will produce:

```python
text = "cat"
token_list = preprocess_fn("<dummy>", {"text": text})["text"]
token_list
# [array([5])]
ground_truth_mat, utt_begin_indices = prepare_token_list(config, text)
ground_truth_mat
# array([[-1],
#        [ 0],
#        [ 5],
#        [ 0]])
```

The previous method scans for tokens:
https://github.com/lumaku/ctc-segmentation/blob/596e20f23dc9a42780b42afeb736a47ace80152f/ctc_segmentation/ctc_segmentation.py#L272-L279

This will produce:

```python
text = ["cat"]
char_list = ["•", "UNK", "a", "c", "t", "cat"]
ground_truth_mat, utt_begin_indices = prepare_text(config, text, char_list)
# array([[-1, -1, -1],
#        [ 0, -1, -1],
#        [ 3, -1, -1],
#        [ 2, -1, -1],
#        [ 4, -1,  5],
#        [ 0, -1, -1]])
```

You can see here that the partial characters are detected (3,2,4), as well as the full "cat" token (5).
This is done to have a better time resolution for the alignment.

Reply:
Another point: The `preprocess_fn` converts unknown chars into the "\<unk>" token.
This may fill up the expected character sequence with "\<unk>" tokens, but these unknown characters should rather be ignored.
Is there an option for `preprocess_fn` to ignore those characters? - or maybe is there already a function to filter out these OOV parts separately?

Reply:
One question: Is the "\<unk>" token included in every espnet 2 dictionary?
https://github.com/espnet/espnet/blob/dab2092bc9c8e184c48cc6e603037333bd97dcd1/espnet2/text/token_id_converter.py#L15

If so, it would be easy to filter the tokenized sequence based on this specific token.

Reply:
> One question: Is the "<unk>" token included in every espnet 2 dictionary?

Yes. 

> char_list = ["•", "UNK", "a", "c", "t", "cat"].

I'm not sure. The token list should be char level or bpe symbols. What do you mean here?

Reply:
OK, I included a filter for the "\<unk>" token.

Reply:
> I'm not sure. The token list should be char level or bpe symbols. What do you mean here?

For example, the char list of a model includes "a", "al", "as", "▁", "▁a", "▁about", "▁actually" ... when configured correctly, the above mentioned method finds parts of the text corresponding to the tokens.

Reply:
<details><summary>Example with a bpe 500 model char list from tedlium 2</summary><div>

```python

from ctc_segmentation import CtcSegmentationParameters
from ctc_segmentation import prepare_text

char_list = [ "<unk>", "'", "a", "ab", "able", "ace", "ach", "ack",
"act","ad","ag", "age", "ain", "al", "alk", "all", "ally", "am",
"ame", "an","and", "ans", "ant", "ap", "ar", "ard", "are",
"art","as", "ase","ass", "ast", "at", "ate", "ated", "ater", "ation",
"ations", "ause","ay", "b", "ber", "ble", "c", "ce", "cent",
"ces","ch", "ci", "ck","co", "ct", "d", "de", "du", "e", "ear",
"ect", "ed", "een", "el","ell", "em", "en", "ence", "ens",
"ent","enty", "ep", "er", "ere","ers", "es", "ess", "est", "et", "f",
"fe", "ff", "g", "ge", "gh","ght", "h", "her", "hing", "ht",
"i","ia", "ial", "ib", "ic","ical", "ice", "ich", "ict", "id", "ide",
"ie", "ies", "if","iff","ig", "ight", "ign", "il", "ild", "ill","im",
"in", "ind","ine", "ing", "ink", "int", "ion", "ions", "ip",
"ir","ire","is","ish", "ist", "it", "ite", "ith", "itt", "ittle",
"ity", "iv","ive", "ix", "iz", "j", "k", "ke", "king", "l",
"ld","le","ll","ly", "m", "ment", "ms", "n", "nd", "nder", "nt", "o",
"od","ody", "og", "ol", "olog", "om", "ome", "on",
"one","ong","oo","ood", "ook", "op", "or", "ore", "orm", "ort",
"ory", "os","ose", "ot", "other", "ou", "ould", "ound",
"ount","our","ous","ousand", "out", "ow", "own", "p", "ph", "ple",
"pp", "pt","q", "qu", "r", "ra", "rain", "re", "reat", "red",
"ree","res","ro","rou", "rough", "round", "ru", "ry", "s", "se",
"sel", "so","st","t", "ter", "th", "ther", "ty", "u", "ually",
"ud","ue", "ul","ult", "um", "un", "und", "ur", "ure", "us", "use",
"ust", "ut","v","ve", "vel", "ven", "ver", "very", "ves", "ving","w",
"way","x", "y", "z", "ăť", "ō", "▁", "▁a", "▁ab",
"▁about","▁ac","▁act","▁actually", "▁ad", "▁af", "▁ag", "▁al",
"▁all", "▁also","▁am", "▁an", "▁and", "▁any", "▁ar",
"▁are","▁around", "▁as","▁at","▁b", "▁back", "▁be", "▁bec",
"▁because", "▁been", "▁being","▁bet", "▁bl", "▁br", "▁bu",
"▁but","▁by", "▁c", "▁call","▁can","▁ch", "▁chan", "▁cl", "▁co",
"▁com", "▁comm", "▁comp","▁con", "▁cont", "▁could", "▁d",
"▁day","▁de", "▁des","▁did","▁diff", "▁differe", "▁different",
"▁dis", "▁do", "▁does","▁don", "▁down", "▁e", "▁en",
"▁even","▁every", "▁ex", "▁exp","▁f","▁fe", "▁fir", "▁first",
"▁five", "▁for", "▁fr", "▁from", "▁g","▁get", "▁go",
"▁going","▁good", "▁got", "▁h", "▁ha","▁had","▁happ", "▁has",
"▁have", "▁he", "▁her", "▁here", "▁his","▁how", "▁hum",
"▁hundred","▁i", "▁ide", "▁if", "▁im", "▁imp","▁in","▁ind", "▁int",
"▁inter", "▁into", "▁is", "▁it", "▁j", "▁just","▁k","▁kind",
"▁kn","▁know", "▁l", "▁le", "▁let", "▁li", "▁life","▁like",
"▁little", "▁lo", "▁look", "▁lot", "▁m",
"▁ma","▁make","▁man","▁many", "▁may", "▁me", "▁mo", "▁more",
"▁most","▁mu", "▁much", "▁my", "▁n", "▁ne", "▁need", "▁new",
"▁no","▁not","▁now", "▁o", "▁of", "▁on", "▁one","▁only", "▁or",
"▁other", "▁our","▁out", "▁over", "▁p", "▁part", "▁pe",
"▁peop","▁people","▁per","▁ph", "▁pl", "▁po", "▁pr", "▁pre", "▁pro",
"▁put", "▁qu", "▁r","▁re", "▁real", "▁really", "▁res",
"▁right","▁ro", "▁s","▁sa","▁said", "▁say", "▁sc", "▁se", "▁see",
"▁sh", "▁she", "▁show", "▁so","▁som", "▁some", "▁somet","▁something",
"▁sp","▁spe", "▁st","▁start", "▁su", "▁sy", "▁t", "▁ta",
"▁take","▁talk", "▁te","▁th","▁than", "▁that", "▁the", "▁their",
"▁them", "▁then", "▁there","▁these", "▁they", "▁thing",
"▁things","▁think", "▁this","▁those","▁thousand", "▁three",
"▁through", "▁tim", "▁time", "▁to", "▁tr","▁tw", "▁two", "▁u",
"▁un","▁under", "▁up", "▁us","▁v", "▁very","▁w", "▁want", "▁was",
"▁way", "▁we", "▁well", "▁were","▁wh","▁what", "▁when",
"▁where","▁which", "▁who", "▁why", "▁will","▁with", "▁wor", "▁work",
"▁world", "▁would", "▁y","▁year","▁years", "▁you", "▁your"]

text = ["I ▁really ▁like ▁technology", "The quick brown fox jumps over the lazy dog.", "unknown chars äüößß-!$ "]
config = CtcSegmentationParameters()
config.char_list = char_list
ground_truth_mat, utt_begin_indices = prepare_text(config, text)

# ground_truth_mat
# array([[ -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  0,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [244,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [190, 410,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 55, 193, 411,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  2,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [137,  13,  -1,  -1, 412,  -1,  -1,  -1,  -1,  -1],
#        [137, 140,  15,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [240, 141,  -1,  16,  -1,  -1, 413,  -1,  -1,  -1],
#        [244,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [137, 356,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 87,  -1, 359,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [134,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 55, 135,  -1,  -1, 361,  -1,  -1,  -1,  -1,  -1],
#        [244,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [209, 438,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 55,  -1, 442,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 43,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 83,  47,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [145,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [137, 153,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 79, 152,  -1, 154,  -1,  -1,  -1,  -1,  -1,  -1],
#        [240,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  0,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 83,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 55,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [188,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [214, 189,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 87,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 43,  91,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [134,  49,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 40,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [190,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149, 198,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [237, 181,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [145,  -1, 182,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 76,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [239,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [133,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [214,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [142, 220,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [183,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [204,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149, 207,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [229,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 55, 230,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [190,  69, 233,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [209,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 83, 211,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 55,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [137,  60,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  2,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [241,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [240,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 52,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 79, 152,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  0,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [214,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [145, 221,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [134,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [145,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [149,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [237, 181,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [145,  -1, 182,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 43,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [ 83,  47,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  2,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [190,  24,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [204,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1],
#        [  0,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1]])
```

</div></details>

Reply:
In the example, parts of the word "▁really" were separated into the token ids: [244, 410, 411, 412, 413]
This corresponds to `['▁', '▁r', '▁re', '▁real', '▁really']`

The CTC segmentation algorithm then iterates over these tokens in the ground truth, calculates the transition probabilities for each token from `lpz` and takes the token combination with the highest combined transition probability.

Reply:
Thanks. I understood.

Then, we need to use some parts of the preprocessor:

https://github.com/espnet/espnet/blob/854afb8c0aa506a2aadc48c6259f64824c506c57/espnet2/train/preprocessor.py#L305-L308

```python
# e.g.
text = preprocress_fn. text_cleaner(text)
tokens = preprocress_fn.tokenizer.text2tokens(text)
(...create grand truth matrix using token_list)
```

At least, we must invoke `text_cleaner` for cleaning since you didn't apply it.

Reply:
Yes. The text cleaner is very helpful. Pre-trained ASR models usually don't have text cleaners - what would be the right way to add new text cleaners? Should I add a `text_cleaner: Collection[str] = None,` parameter to the `CTCSegmentation` object?

Reply:
Do you need text_cleaner for CTCSegmentation now? Currently it's not used for all asr recipe now as you said.

Reply:
OK - For now, let's keep the already included cleaners. It's still possible to add a parameter for new cleaners later...

Text cleaning is certainly an essential step. That's what we did with the spacy framework in the [CTC segmentation paper](https://arxiv.org/abs/2007.09127), e.g., changing numbers to words, changing orthography, etc.

It is more important for establishing a good ground truth than for the actual alignment. Therefore, I would rather expect that text cleaning and text normalizing happens as a separate step before doing the segmentation. Doing text normalization as a separate step allows to manually check whether the result if acceptable.

Reply:
Sorry for late comment.

Could you change the code to my idea and remove the classic part?
Sorry, I'm not sure I could tell you my intention in the past comment.


```python
# e.g.
text = preprocress_fn. text_cleaner(text)
tokens = preprocress_fn.tokenizer.text2tokens(text)
(...create grand truth matrix using token_list)
```

Reply:
Sorry, I misunderstood your comment because I thought that `preprocess_fn.text_cleaner()` was already part of `preprocess_fn()`.
Changed it!

And also, for the classic method, true, it is used only in certain ASR recipes.
It is useful in those situations with long tokens that can be broken down in partial tokens.
I removed the option and left instructions in the form of a docstring for advanced users.

Reply:
Sorry, I meant please remove not classic one and merge into one, classic only.
Then, I think , you need character/token lists instead of numpy.ndarray, so I thought we can't use `preprocess_fn()` because it converts it to ndarray.   We need to use `preprocress_fn. text_cleaner` and `preprocress_fn. text2tokens` in preprocress_fn.

Reply:
Also note that preprocess_fn includes text_cleaner(): https://github.com/espnet/espnet/blob/fee8f1755e91d01bbefae4904630c3653fb1a379/espnet2/train/preprocessor.py#L304-L308

Reply:
Ah - now I understand better.
It's a little bit complicated - using the "classic" method has a disadvantage and an advantage.

The advantage is that it detects partial tokens.
Howeever, I think that this has also one disadvantage:
To explain this, I need to go into the details of how the `prepare_text` function is implemented.

Your proposed idea is great and correct, it helps the "classic" to better find tokens because of the preprocessing,
the implementation would maybe look like this:
```python
text = [self.preprocess_fn.text_cleaner(utt) for utt in text]
token_list = [" ".join(self.preprocress_fn.tokenizer.text2tokens(utt)) for utt in text]
token_list = [utt.replace("<unk>", "") for utt in token_list]
ground_truth_mat, utt_begin_indices = prepare_text(config, token_list)
```
The output would be something like `This <space> sequence <space> has <space> tokens <space> as <space> spaces.`
With your proposed method, all tokens will be detected correctly.

But when we compare the string-length of the ground truth:
```python
len("This sequence has tokens as spaces.")
# 35
len("This<space>sequence<space>has<space>tokens<space>as<space>spaces.")
# 56
```
As you can see the `token` sequence is longer.
The `prepare_text` function generates a ground truth matrix with dimensions `[length of the ground truth, length of longest token]`.
With this, in some cases, it happens that the prepared `ground_truth_mat` is longer than time steps in lpz (`lpz.shape[0]`)
and this causes an error in `ctc_segmentation`.
https://github.com/lumaku/ctc-segmentation/blob/905d6cc066cd7380b9d023050f5aa9b895eace0c/ctc_segmentation/ctc_segmentation.py#L142-L143
I would estimate, on average one token activation every 4 lpz indices, therefore, this error is not unprobable.

As the "easier" solution, I set the ESPnet 2 "tokenize" as default.
I argue that the "tokenize" method also prevents one error specific to the method how espnet 2 tokenizes.

I would argue: In 70% of use cases, the regular `tokenize` method is the best.
In the other 30% of cases, with the long tokens, then it would be be better to use the "classic" method.
Therefore, both methods have a reason.

You probably will prefer only one method to reduce the maintenance overhead.
For this, I recommend the "tokenize" method...

Reply:
The "tokenize" method is more efficient and the easiest:
* For models based on latin / cyrillic script that only contains the main characters, `["a", "b", ...]`, the "tokenize" method is already.
* For a Japanese or Chinese ASR model with e.g. ~3000 tokens Kanji / Hanzi

For models that contain long tokens, such as `_really`, the "classic" method improves the accuracy of the alignments, but with a greater complexity for implementation and computation...

Reply:
> It's a little bit complicated - using the "classic" method has a disadvantage and an advantage.

Oh, I see.

> You probably will prefer only one method to reduce the maintenance overhead.

No, I misunderstood that classic one is better. My intension is using preprocess_fn for tokenization, so I think we should have two methods .

Reply:
Yes, I agree. I reverted to the commit with both methods and added documentation on `text_converter`. Thanks!

Reply:
Please use the follwing in classic. We need to use text2tokens. My intention is here. External tools can't know how to tokenize the text.

```
text = [self.preprocess_fn.text_cleaner(utt) for utt in text]
token_list = [" ".join(self.preprocress_fn.tokenizer.text2tokens(utt)) for utt in text]
token_list = [utt.replace("<unk>", "") for utt in token_list]
ground_truth_mat, utt_begin_indices = prepare_text(config, token_list)
```

Reply:
I want to prevent the issue with too long ground truth sequences caused by long tokens, primarily by the space token `<space>`.
I think that a possible solution is to replace all instances of this token by space, i.e. to change `<space>` to " " at the tokenizer (`aligner.preprocess_fn.tokenizer.space_symbol`). Would this work?

Reply:
Sorry, I didn't understand yet. How do you handle the problem in the current classic way? What is the difference?

Reply:
In the Espnet 1, the space token was just " " or it was ignored.
The difference is the length of the ground truth, see the example above: See `len("This sequence has tokens as spaces.")` versus `len("This<space>sequence<space>has<space>tokens<space>as<space>spaces.")`.

Reply:
Why not be it done now as same? I think this situation is completely same.

Reply:
There is a small difference in how the `prepare_text` handles text.
But it's not that important. Sorry for the confusion.
The latest commit does not handle the space token differently.
Please check - If you are OK with this, then we can merge

Reply:
If you are okay, it's okay for me, but I couldn't understand why don't you do as same? It seems easy.

Reply:
Ha! You are right, why make it easy if we can make it complicated :wink: ?
I am OK with the code as it is.

Reply:
The CI test failed - it seems because of an unknown error. Can we restart it?